### PR TITLE
Rename `System` to `Backend`

### PR DESCRIPTION
**Pull request by bveeramani**

### CODE REVIEW

**Review by elibol — COMMENTED** [review]
Looks good! Left some minor comments.

Another refactor to consider (after this is merged): `nums.core.compute` (and everything "compute" related) to `nums.core.kernel`. e.g. `ComputeInterface` would become `KernelInterface`, etc. Kernel is a bit more common and closer to the actual semantics of these modules and classes.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -23,9 +23,9 @@ or `nums.core.array.blockarray.BlockArray`.
 if it's not possible to implement using existing methods, 
 or the implementation's execution speed can be 
 improved beyond what is achievable using existing methods.
-4. Add kernel interfaces to `nums.core.systems.interfaces.ComputeInterface`, and implement
+4. Add kernel interfaces to `nums.core.backends.interfaces.ComputeInterface`, and implement
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply CONTRIBUTING.md:26]: This was an issue before, and is unrelated to this PR. I'll fix it in a follow-up PR.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -23,9 +23,9 @@ or `nums.core.array.blockarray.BlockArray`.
...
 the interface methods for all existing compute implementations.
-Currently, the only compute interface is `nums.core.systems.numpy_compute`.
+Currently, the only compute interface is `nums.core.backends.numpy_compute`.
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply CONTRIBUTING.md:28]: This was an issue before, and is unrelated to this PR. I'll fix it in a follow-up PR.

File: nums/core/application_manager.py [hunk 3]
```diff
@@ -23,9 +23,13 @@
 from nums.core.compute import numpy_compute
 from nums.core.compute.compute_manager import ComputeManager
 from nums.core.grid.grid import DeviceGrid, CyclicDeviceGrid, PackedDeviceGrid
-from nums.core.systems.filesystem import FileSystem
-from nums.core.systems.system_interface import SystemInterface
-from nums.core.systems.systems import SerialSystem, RaySystem, RaySystemStockScheduler
+from nums.core.backends.filesystem import FileSystem
+from nums.core.backends.backend_interface import Backend
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply nums/core/application_manager.py:27]: There's no wide-spread convention for naming abstract base classes. 

Anecdotally, I don't think base classes are usually prefixed or postfixed in anyway (e.g., `Backend`).  The built-in `collections.abc` package doesn't prefix or postfix base class names in anyway. Neither does Torch. 

If there is a prefix convention I'd recommend, it's `Base` (e.g, `BaseBackend`). sklearn follows this convention for base classes.

I strongly oppose `IBackend` and moderately oppose `BackendInterface` as those seem un-Pythonic.

My suggestion is to stick with no prefix or postfix (e.g., keep `Backend`). W.r.t to the issue of consistency, there are only two other interfaces (`ComputeInterface` and `RNGInterface`), so we can always rename those to adhere to whatever convention we decide.

**Review by elibol — APPROVED** [review]
Would like to revisit base/interface class naming conventions again in the future. Other than that, looks good to me 🙂

**Pull request merged**